### DISCUSSION TITLE
General

### DISCUSSION THREAD

**Discussion by Try** [Try/OpenGothic#106] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Try** [comment]
For all contributors:

Submodule `MoltenTempest` was renamed to just `Tempest`, so be aware!

**Comment by TomislavTom** [comment]
Has it become more challenging to add new functions or achieve new breakthroughs from your ToDo list?
I'm asking because there was (unusually) no release for a long time and new videos are also not coming as frequent as they used to come.
However I can see that you're commiting new changes almost daily but many seem to be just refactorings...
Is there a plan to add SSAO any time soon?
I think SSAO and an adjusted mouse rotation would definitely make this project go viral in the Gothic community.
The question is just if you would want it to go viral at this stage of development.
Keep up the good work!

**Reply by Try** [reply]
Hi, @TheDovahkiin97 !

Actually I have a plans to publish release a few week ago, it just keep get delayed.
Main problem with this release is: I have almost working physic, for dropped items; almost ready spell/effects system; almost ready global effects, but nothing fine polished.  

Hm, you know what? - I'll publish build today, it's stable enough and then continue work on finalizing stuff.

**Reply by Try** [reply]
> Is there a plan to add SSAO any time soon?

Yes, I even have a few more ideas how to make a better visual:
1. Better shadow-maps (just tune-up current implementation), fixing light-leaks on edges.
2. Add some global illumination shader, can be SSAO or SSDO
3. Combine №2 with image-based light

**Reply by TomislavTom** [reply]
Wow I didnt assume that asking would be enough for an instant release haha
All of this sounds very good but dont put extra pressure on yourself.
Do you think that all the base game functionalities could be finished by the end of Q1/2021?

**Reply by Try** [reply]
> Do you think that all the base game functionalities could be finished by the end of Q1/2021?

I'm intent to finish base functionality in Q1, but may be it will be delayed until Q2: face animation and effect are tough.

**Comment by Try** [comment]
New binary build published:
https://github.com/Try/OpenGothic/releases/tag/gothic-win32-v1.0.908

**Comment by BlackMagicCoding** [comment]
I don't know if this is the right place here, but I had a question ever since I saw your project:
Will this only be able to run Gothic 2, or Gothic 1 as well?
Afaik both of those games used the same engine, but I am not too familiar with the very details.
Nevertheless, after the death of ReGoth I am happy that someone made a new project for this absolute gems of a game, thank you!

**Reply by Try** [reply]
Hi, @LifeIsACurse !

The project is aiming only for Gothic2/gothic2-notr. You can try to run Gothic1, but game experience won't be nice - there are a lot of minor differences between G1 and G2 engine.

**Reply by Matriks404** [reply]
But, is Gothic 1 support planned for the future, or is it something you'd not plan doing at all?

**Reply by Try** [reply]
Plan so far: finalize gothic2 support, make some visual improvements and then will see.

**Comment by progward** [comment]
Judging from your Todo-List, amazed to see you've come so far already. It almost looks like you're almost finished with the project. How much percent do you think is left to finish it? I know that's usually not easy to answer, but what's your feeling?

**Reply by Try** [reply]
Well, last time I was asked about it, the response was: early 2021 :D
On the other hand there is only one mission-critical issue remain: AOE spells, like rain, ice-wave and such.

**Comment by BlackMagicCoding** [comment]
Hello @Try - would you consider writing a little summary how this whole thing functions in theory?
I am really interested how it came you wanted to do this project and what was/is the path you decided on to solve the problem.
Was it "just" reverse engineering the original binaries and re-implementing the functions in a new engine?
Kind regards and thanks for your great work :)

**Reply by Try** [reply]
HI, @LifeIsACurse !

Two main reasons for this project:
  * original Gothic II works very poorly on my machine
  * this is a good real-usage case for my graphics engine
 
About implementation approach: I used ZenLib to deal with the asset-loading; and game logic is quite straightforward: all I have to do is to implement all script-api entry's, so scripts from original game can work the same way as they were intended.

**Reply by progward** [reply]
I've seen somewhere that all the Mods will probably work with your engine. Are you aware of GMP software, and do you think it might work together with your engine as well? It places hooks in the original game so I imagine you would have to adapt to it then

**Reply by Try** [reply]
I'm aware of AST-sdk and Union, wasn't aware about GMP.

Had a discussion with Union dev's - no way we can make it compatible, because internal structure in OpenGothic is very different that it is in original game. Most of the time there is nothing to hook into, on my side. 

As alternative I can provide a extension set to script-api, so mod makers can have access to 2d-drawing subsystem and such, but it won't do anything for already existing content.

**Reply by pkurras** [reply]
@Try this is not 100% correct. Gmp works with messages internally. If it is integrated into the project properly there would be no need for hooking. Just send and receive the correct message id's and do with them what your engine should do. It's like a big message system. Should be easy enough to do.

**Reply by progward** [reply]
That'd be awesome.

**Comment by pm4rcin** [comment]
The game runs great but the main problem is the camera control. With keyboard it's not that bad but with mouse it's really weird to control it. Are there any plans to change it to make it better?

**Reply by Try** [reply]
Camera behavior in original game is quite complex sometimes and hard to mimic it.
Probably current camera is the best we can have. 
There is a camera-ticket here: https://github.com/Try/OpenGothic/issues/101

Btw, assuming you use arch-distro: do you know what build they providing? Since they do it by them selfs, there is a possibility that their build is just very old one.

**Reply by pm4rcin** [reply]
It was built from 0.37 tag. Submodules were built from fresh commits at the time of release. You can easily check it in [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=opengothic) And yes I use Arch btw. ;)

**Comment by pm4rcin** [comment]
Since mods should work normally how one can select to play different mod?

**Reply by Try** [reply]
So far I've tested only gothic reload mod - just installed in on top of the game, same way as for vanilla.
If you looking for something, like GothicModStarter - it's not there.

**Comment by TomislavTom** [comment]
Hi,
are you perhaps working on a new video?
Also it would be cool if you could update/replace your TODO list.
From what I understood only some spells and camera controls are left to make the game playable/finishable, right?

However obviously you still work on other things that are not on the list so im wondering what the next priorities are or what you generally want to do with this project in the future.

Anyways good job and keep it up! 👍

Also if anyone runs this on some better hardware i would like to hear how it performs in terms of fps/resolution.

**Reply by Try** [reply]
Hi, @TheDovahkiin97 !
>perhaps working on a new video?

Well, not really. My short-term plan is to finalise MacOS/Metal support and then make a big review on graphics api in engine. 
Apart from that, thinking about special video about visual effects and spells, once those are done.

> Also if anyone runs this on some better hardware

Now running on M1/Retina: have stable 60 outside of city, in city ~40. In city case seem to be bound on deferred shading, but will see 
Here: 2880x1800
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/122894/118393183-069b4a00-b63e-11eb-901d-b7c7f67607f4.png">

**Reply by clort81** [reply]
> Now running on M1/Retina: have stable 60 outside of city, in city ~40. In city case seem to be bound on deferred shading, but will see

Beautiflu!
I'm part of that poor (Arm) group using linux arm SBCs.  The wonderful people behind Panfrost have given us Opengl and GL-es (<=3.0) on these.  The $100 class of SBCs with best GPUs are currently Odroid N2+ and Khadas Vim3 Pro.  These have the Mali G52 gpu.

Panfrost Vulkan support is WIP but some things work.

The heavier games that run reasonably are Super TuxKart and Doom3.  OpenMW with low draw distance is playable as well.  Super TuxKart manages around 120-130k polys at 20-30fps if shaders are limited.  Glmark2-es2-wayland gives us a score around 900.

If you could keep in mind ways to deliver usable FPS to a generation below the M1 gpu, we'd be super grateful.

The current work on Panfrost Vulkan can be followed here [https://oftc.irclog.whitequark.org/panfrost/2021-07-09](https://oftc.irclog.whitequark.org/panfrost/2021-07-09)
All the best.

**Comment by TheEpicNoobZilla** [comment]
Do i need Vulkan capable GPU to use it?

**Reply by Try** [reply]
Basically yes: you gpu has to support Vulkan 1.0 or DirectX12 or Metal(on MacOS)

**Reply by TheEpicNoobZilla** [reply]
Damn, so i will stick to wine with Gallium Nine

**Comment by Fabiosek** [comment]
Hi. Will you improve optimization in the future? 20-40 fps is not a good result.

**Reply by Try** [reply]
Hi, @Fabiosek !
What is your system setup? I have ~30 in debug and 60 in release on my laptop, when running vanilla content.

**Reply by Fabiosek** [reply]
> 
> 
> Hi, @Fabiosek !
> What is your system setup? I have ~30 in debug and 60 in release on my laptop, when running vanilla content.

Windows 10 PRO 64BIT:
1. i5-4670S, r9 380X 4GB, 8GB RAM, SSD (40 fps in city, 60fps outdoor)
2. i5-7600K, Geforce GTX 1060 6GB, 8GB RAM, SSD (40-54 fps outdoor)

PS. Gothic is an old game. Even new equipment will not give you more fps. Currently, the only solution is "DX11-mod". Your engine rewrite works on "newer hardware" but fps is still 30-40. It should always be 60 fps, It should still be a constant number of fps.  

PS2. I still have a laptop with an "i5" processor. I think the CPU usage is too high and / or it's a problem with "intel" (maybe i5)?

**Reply by Try** [reply]
On CPU performance side one thing is in plan: make scene-graph structure, to discard objects not by one, but by group instead.
On GPU side - little room for any significant improvement, without bumping system requirements(from Vulkan 1.0 to 1.2)

PS. Gothic is old game - true; yep OpenGothic is not the same game: "true" world simulation, bullet physics and extra vision range has effect.

PS2. Here is CPU breakdown:
![изображение](https://user-images.githubusercontent.com/122894/124175628-1d8ee200-daae-11eb-9d5f-e5c893d5597b.png)

Main takeaways are: 
1. `ObjectsBucket::drawCommon = 36%` - command recording.
2. `Gothic::tick = 37%` - npc(mostly) processing. 17% of it is collision detection. (BulletPhysics here)
3. `WorldObjects::updateAnimation = 9%` - animation 
4. Sound - 4% (out of screenshot); Swapchain = 4%
5. Remaining (~10%) - unsorted

GPU breakdown:
![изображение](https://user-images.githubusercontent.com/122894/124176813-b40fd300-daaf-11eb-910b-0d8b6c018d7b.png)

55.8k drawcalls - Gothic is full of small objects, without much of info of world structure to effectively manage them.

**Comment by LucyJohnson-lua4life** [comment]
General question to all people following this project: is there any team known thats planing on developing a multiplayer around Open Gothic? 

Would be really great if one could access the new clean engine api for a multiplayer project.

**Reply by Try** [reply]
Hi, @LucyJohnson-lua4life !

From what I know, there is no active work on network at this moment.  Also I have no clear vision on what kind of api mod-developer want to have for this case.
So any contribution is very welcome :)

**Reply by progward** [reply]
An API for the client would be a start, for letting the game do stuff you want. Like playing animations, triggering mobs, setting a players position, setting the player's attributes. All this stuff. Here's a list of functions that would especially be important for a Multiplayer project (minus all the server-related functions of course).

I think this would give you a good idea.

https://gitlab.com/Reveares/GMP/-/wikis/Functions

**Reply by progward** [reply]
Here's a list of another multiplayer project. Here it's mostly only client-side functions, so you don't have to filter out that much. :)
https://gothic-online.com.pl/wiki/Script:Client_Functions

**Reply by clort81** [reply]
> General question to all people following this project: is there any team known thats planing on developing a multiplayer around Open Gothic?

Creating a multiplayer game - even on top of a working engine - involves vast amounts of work.  If you seriously wish to contribute to one, I would suggest looking at Eternal Lands, TES3MP (Morrowind Multiplayer), TheManaworld or Daimonin projects.
